# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Zone.h
# Commit: 5a6684922b98
# Full Hash: 5a6684922b980d32b05bd7210cfadc34e734a845
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2020-03-21 21:28:09
# Regressor Bug: 1623973
# File Overlap Count: 2
# Description:
#   Bug 1623973 - Make FinalizationRegistryObject active record set and registrations map values weak r=sfink
#   
#   This removes tracing for the values of the registration map and the active record set and sweeps them instead. This requires maintaining a map of FinalizationRegistryObjects on the zone.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D67677
# ==============================================================================

diff -r c4d2ca8f78b7 -r 5a6684922b98 js/src/gc/Zone.h
--- a/js/src/gc/Zone.h	Thu Mar 19 16:42:42 2020 +0000
+++ b/js/src/gc/Zone.h	Fri Mar 20 21:57:22 2020 +0000
@@ -319,6 +319,12 @@
       js::Vector<js::AccessorShape*, 0, js::SystemAllocPolicy>;
   js::ZoneData<NurseryShapeVector> nurseryShapes_;
 
+  // The set of all finalization registries in this zone.
+  using FinalizationRegistrySet =
+      GCHashSet<js::HeapPtrObject, js::MovableCellHasher<js::HeapPtrObject>,
+                js::ZoneAllocPolicy>;
+  js::ZoneOrGCTaskData<FinalizationRegistrySet> finalizationRegistries_;
+
   // A map from finalization registry targets to a list of finalization records
   // representing registries that the target is registered with and their
   // associated held values.
@@ -685,6 +691,10 @@
 
   void sweepWeakKeysAfterMinorGC();
 
+  FinalizationRegistrySet& finalizationRegistries() {
+    return finalizationRegistries_.ref();
+  }
+
   FinalizationRecordMap& finalizationRecordMap() {
     return finalizationRecordMap_.ref();
   }
